### DISCUSSION TITLE
[Share] Copy DOI to Clipboard

### DISCUSSION THREAD

**Discussion by rodrigolourencofarinha** [windingwind/zotero-actions-tags#397] [Action Scripts]
### Description

This script extracts DOIs from the currently selected Zotero items, formats them with a comma after each DOI (except the last), and copies the resulting list to the clipboard for easy use in other applications.

### Event

None

### Operation

Script

### Data

```javascript
// Copy DOIs with a comma after each, except the last
// Based on copy tags

var doiArray = [];
var items = Zotero.getActiveZoteroPane().getSelectedItems();
for (let item of items) {
    if (item.isRegularItem() && !(item instanceof Zotero.Collection)) {
        let doi = item.getField('DOI');
        if (doi && !doiArray.includes(doi)) {
            doiArray.push(doi);
        }
    }
}

// Create a string with a comma after each DOI except the last one
var clipboardText = doiArray.map((doi, index) => {
    return index < doiArray.length - 1 ? doi + ',' : doi;
}).join('\r\n');

Zotero.Utilities.Internal.copyTextToClipboard(clipboardText);
return "DOIs copied";
```

**Comment by windingwind** [comment]
https://github.com/windingwind/zotero-actions-tags?tab=readme-ov-file#custom-script

Please always use the items instead of ZoteroPane.getSelectedItems(). The action can be triggered from entrances outside the library, where the items are not from ZoteroPane.

**Comment by loune20** [comment]
The script above does not work for me (if I select one item, the DOI is visible in my clipboard history, if I select several items nothing is visible in the history ; in any case doing `Ctrl+V` doesn't work).
Any pointers would be appreciated, my code below.

I am using Zotero `9` and Action & Tags `2.5.1` on Ubuntu if that's relevant.

```js
var doiArray = [];
for (let item of items) {
    if (item.isRegularItem() && !(item instanceof Zotero.Collection)) {
        let doi = item.getField('DOI');
        if (doi && !doiArray.includes(doi)) {
            doiArray.push('https://doi.org/'+doi);
        }
    }
}

// Create a string with a comma after each DOI except the last one
var clipboardText = doiArray.map((doi, index) => {
    return index < doiArray.length - 1 ? doi + ',' : doi;
}).join('\r\n');

Zotero.Utilities.Internal.copyTextToClipboard(clipboardText);

// Tried this way, didn't work either
// const clipboard = new Zotero.ActionsTags.api.utils.ClipboardHelper();
//clipboard.addText(clipboardText);
//clipboard.copy();

// This does return the correct DOI(s), so the issue seems to be with copying to clipboard
return clipboardText;
```

**Reply by jcblum** [reply]
I think the problem you're seeing is probably not related to the clipboard part of the code, but instead may be because the script is based on `items`, and therefore needs to include a line like this right at the top:

```javascript
// If the `item` variable is truthy, then skip running since
// this is a per-item execution and the `items` variable will be empty.
if (item) return;
```

This is a tricky aspect of using `items`. From [the docs](https://github.com/windingwind/zotero-actions-tags#custom-script):

> When selecting multiple items in the library, the action will be triggered once for all items (`items`=`[...]`, `item`=`undefined`) and then one by one for each item (`items`=`[]`, `item`=`...`). You can use the `items` variable to get the selected items array and avoid duplicate executions.

The key part is that the batch operation run (when the built-in `items` variable is non-empty, but the `item` variable is undefined) happens *first*. If you don't include a gate as shown above (or in the `items` examples), then the per-item runs occur immediately afterward. During those runs, `items` still exists, but is an empty array. That means all the `items`-based code in the script runs again on the empty `items` array, overwriting the earlier output. 

Although the wording in the docs made me initially think that the duplicate execution only happens when you have multiple items selected, I can confirm that on my system (Zotero 9.0.4, Actions & Tags 2.5.2), both the batch and per-item runs occur even when only one item is selected, as long as the script is triggered from a keyboard shortcut or a menu. So you should always include the `if (item) return;` gate any time you're basing your script on `items`.

**Comment by jcblum** [comment]
I wrote a script aiming at the same goal, but it wound up diverging substantially enough that I've posted it as a new [Share] (#624).
